src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Windows.cs

```diff
@@ -525,6 +524,106 @@ private unsafe SortKey CreateSortKey(string source, CompareOptions options)
             return new SortKey(this, source, options, keyData);
         }
 
+        private unsafe int GetSortKeyCore(ReadOnlySpan<char> source, Span<byte> sortKey, CompareOptions options)
+        {
+            Debug.Assert(!GlobalizationMode.Invariant);
+            Debug.Assert((options & ValidCompareMaskOffFlags) == 0);
+
+            // LCMapStringEx doesn't allow cchDest = 0 unless we're trying to query
+            // the total number of bytes necessary.
+
+            if (sortKey.IsEmpty)
+            {
+                throw new ArgumentException(
+                    paramName: nameof(sortKey),
+                    message: SR.Argument_CannotBeEmptySpan);
+            }
+
+            uint flags = LCMAP_SORTKEY | (uint)GetNativeCompareFlags(options);
+
+            // LCMapStringEx doesn't support passing cchSrc = 0, so if given an empty span
+            // we'll instead normalize to a null-terminated empty string and pass -1 as
+            // the length to indicate that the implicit null terminator should be used.
+
+            int sourceLength = source.Length;
+            if (sourceLength == 0)
+            {
+                source = string.Empty;
```

Comment:
did you try to check if Windows always return the same sort key for the empty string? so we can avoid calling the OS anyway? I am still going through the commits so I don't know yet if you already did so :-)

Reply:
AFAIK Windows will always return the same sort key for the empty string. But the output value might change between different versions of Windows. So we could cache it at app startup, I suppose.

Do you think getting a _linguistic-aware_ hash code of an empty string is common enough that we should optimize for it?

Reply:
> Do you think getting a linguistic-aware hash code of an empty string is common enough that we should optimize for it?

I don't have data for that but looking at legacy code, most of the time we special case the empty strings which indicate it is good to optimize for. Also, we already checking for empty string anyway, so would be good to do it anyway. you can do that later if you want and don't have to block this work.

Reply:
I think empty strings were special-cased not as an optimization, but because `LCMapStringEx` doesn't correctly handle _cchSrc = 0_.

Reply:
I saw before in the full framework people complained about combat issues when using an empty string which raises the question is someone cares about it. You may be right it is not for perf reason but I cannot judge it more without data. if you think it is not worth it to do it now, that is fine.